File: Disassembler.cpp
#include <fstream>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

using namespace std;

struct Instruction {
    string mnem;
    unsigned int size;

    Instruction(const string& m, unsigned int s);
};

Instruction::Instruction(const string& m, unsigned int s): mnem(m), size(s) {};

class Disassembler {
private:
    vector<char> data;
    unsigned int index;
    size_t end;
    unsigned int digits;

    string output1Byte(const Instruction& instr) const;
    string output2Byte(const Instruction& instr, unsigned char operand) const;
    string output3Byte(const Instruction& instr, unsigned char operand[2]) const;
public:
    Disassembler(char* filename);
    void run();
};

/*
 * JSON array of arrays from file.
 * Each instruction is represented as an array, like:
 *   ["mvi    c,",  2]
 * where the first element is the mnemonic and the second is the size of the 
 * instruction in bytes.
 *
 * The array indices correspond to the opcodes of the Intel 8080 (i.e. 0x00 
 * through 0xff)
 */
vector<Instruction> getInstructions() {
    ifstream f("instructions.json");
    if (!f.is_open()) {
        cout << "Instructions file not found!" << endl;
        exit(0);
    }
   
    vector<Instruction> table;
    string line;
    while (getline(f, line)) {
        // Not a good implementation, but doesn't require other dependencies
        if (line.size() < 23) {
            continue;
        }

        const size_t first = line.find("\"") + 1;
        const size_t last = line.find("\",");
        const string mnem = line.substr(first, last - first);
        const unsigned int size = stoi(line.substr(20, 1));

        const Instruction instr(mnem, size);
        table.push_back(instr);
    }

    f.close();

    return table;
}

Disassembler::Disassembler(char* filename): index(0), digits(4) {
    ifstream f(filename, ios::binary);
    if (!f.is_open()) {
        cout << filename << " is not a file!" << endl;
        exit(0);
    }

    data = vector<char>(istreambuf_iterator<char>(f), 
        istreambuf_iterator<char>());

    f.close();

    end = data.size();
}

string Disassembler::output1Byte(const Instruction& instr) const {
    const int size = 22;
    char buffer[size];
    snprintf(buffer, size, "%04x %s", index, instr.mnem.c_str());
    return string(buffer);
}

string Disassembler::output2Byte(const Instruction& instr, 
        unsigned char operand) const {
    const int size = 22;
    char buffer[size];

    if (instr.mnem.find("out") != string::npos || 
        instr.mnem.find("in") != string::npos) {
        snprintf(buffer, size, "%04x %s$%02x", index, instr.mnem.c_str(), 
            operand);
    } else {
        snprintf(buffer, size, "%04x %s#%02x", index, instr.mnem.c_str(), 
            operand);
    }

    return string(buffer);
}

string Disassembler::output3Byte(const Instruction& instr, 
        unsigned char operand[2]) const {
    const int size = 22;
    char buffer[size];

    if (instr.mnem.find("lxi") != string::npos) {
        snprintf(buffer, size, "%04x %s#%02x%02x", index, instr.mnem.c_str(), 
            operand[0], operand[1]);
    } else {
        snprintf(buffer, size, "%04x %s$%02x%02x", index, instr.mnem.c_str(), 
            operand[0], operand[1]);
    }

    return string(buffer);
}

const vector<Instruction> instructionTable = getInstructions();

void Disassembler::run() {
    while (index < end) {
        const unsigned char byte = data[index];
        const Instruction instr = instructionTable[byte];
        string msg;

        if (instr.size == 1) {
            msg = output1Byte(instr);
        } else if (instr.size == 2) {
            unsigned char operand = data[index + 1];
            msg = output2Byte(instr, operand);
        } else if (instr.size == 3) {
            unsigned char operand[2];
            operand[0] = data[index + 2];
            operand[1] = data[index + 1];
            msg = output3Byte(instr, operand);
        }

        index += instr.size;

        cout << msg << endl;
    }
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cout << "Usage: ./Disassembler [filename]" << endl;
        exit(0);
    }

    char* filename = argv[1];

    Disassembler d(filename);
    d.run();

    return 0;
}

File: Disassembler.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.logging.Handler;
import java.util.logging.FileHandler;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

class Disassembler {
    private static final Logger logger = createLogger();
    private static final Instruction[] INSTRUCTION_TABLE = getInstructions();
    private byte[] data;
    private int index;
    private int end;
    private int digits;

    private static class Instruction {
        public final String mnem;
        public final int size;

        public Instruction(String m, int s) {
            mnem = m;
            size = s;
        }
    }

    private static Logger createLogger() {
        Handler fh = null;

        try {
            fh = new FileHandler("logs/Disassembler.java.log");
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(0);
        }

        Logger logger = Logger.getLogger(Disassembler.class.getName());
        logger.setUseParentHandlers(false);
        SimpleFormatter formatter = new SimpleFormatter();
        fh.setFormatter(formatter);
        logger.addHandler(fh);
        
        return logger;
    }

    /*
     * JSON array of arrays from file.
     * Each instruction is represented as an array, like:
     *   ["mvi    c,",  2]
     * where the first element is the mnemonic and the second is the size of 
     * the instruction in bytes.
     *
     * The array indices correspond to the opcodes of the Intel 8080 (i.e. 0x00 
     * through 0xff)
     */
    private static Instruction[] getInstructions() {
        BufferedReader br = null;
        Instruction[] table = new Instruction[256];
        int count = 0;
        String line = null;

        try {
            br = new BufferedReader(new FileReader("instructions.json"));

            while (null != (line = br.readLine())) {
                // Not a good implementation, but doesn't require other 
                // dependencies
                if (line.length() < 23) {
                    continue;
                }

                final int first = line.indexOf("\"") + 1;
                final int last = line.indexOf("\",");
                final String mnem = line.substring(first, last);
                final int size = Character.getNumericValue(line.charAt(20));

                final Instruction instr = new Instruction(mnem, size);
                table[count] = instr;
                count++;
            }
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(0);
        } finally {
            try {
                if (null != br) {
                    br.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        return table;
    }

    public Disassembler(String filename) {
        Path p = FileSystems.getDefault().getPath("", filename);
        try {
            data = Files.readAllBytes(p);
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(0);
        }
        index = 0;
        end = data.length;
        digits = String.valueOf(end).length();
    }

    private String output1Byte(final Instruction instr) {
        return String.format("%04x %s", index, instr.mnem);
    }

    private String output2Byte(final Instruction instr, final int operand) {
        if (instr.mnem.contains("out") || instr.mnem.contains("in")) {
            return String.format("%04x %s$%02x", index, instr.mnem, operand);
        }

        return String.format("%04x %s#%02x", index, instr.mnem, operand);
    }

    private String output3Byte(final Instruction instr, final int[] operand) {
        if (instr.mnem.contains("lxi")) {
            return String.format("%04x %s#%02x%02x", index, instr.mnem, 
                operand[0], operand[1]);
        }

        return String.format("%04x %s$%02x%02x", index, instr.mnem, operand[0], 
            operand[1]);
    }

    public void run() {
        while (index < end) {
            // Convert signed byte 
            final int b = data[index] & 0xff;
            final Instruction instr = INSTRUCTION_TABLE[b];
            String msg = null;

            if (instr.size == 1) {
                msg = output1Byte(instr);
            } else if (instr.size == 2) {
                int operand = data[index + 1] & 0xff;
                msg = output2Byte(instr, operand);
            } else if (instr.size == 3) {
                int[] operand = new int[2];
                operand[0] = data[index + 2] & 0xff;
                operand[1] = data[index + 1] & 0xff;
                msg = output3Byte(instr, operand);
            }

            index += instr.size;

            System.out.println(msg);
            logger.log(Level.FINE, msg);
        }
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java Disassembler [filename]");
            System.exit(0);
        }

        String filename = args[0];

        Disassembler d = new Disassembler(filename);
        d.run();
    }
}

